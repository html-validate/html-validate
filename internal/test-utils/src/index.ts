export {
	type Source,
	type Transformer,
	type TransformerResult,
	type TransformerChainedResult,
} from "html-validate";
export { transformFile, transformSource, transformString } from "./test-utils";
