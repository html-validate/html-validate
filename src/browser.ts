/* used when calling require('htmlvalidate'); */

export { default as HtmlValidate } from "./htmlvalidate";
export { AttributeData } from "./parser";
export { Config, ConfigData, ConfigError, ConfigLoader, Severity, configPresets } from "./config";
export { DynamicValue, HtmlElement, NodeClosed, TextNode } from "./dom";
export { EventDump, TokenDump } from "./engine";
export { UserError, SchemaValidationError } from "./error";
export * from "./event";
export { MetaData, MetaElement, MetaTable, MetaCopyableProperty } from "./meta";
export { Rule, RuleDocumentation } from "./rule";
export { Source, Location, ProcessElementContext } from "./context";
export { Report, Reporter, Message, Result } from "./reporter";
export { TransformContext, Transformer, TemplateExtractor } from "./transform";
export { Plugin } from "./plugin";
export { Parser } from "./parser";
export { ruleExists } from "./utils";

/* Bug #127: this line need special care, see rollup.config.js */
export { version } from "../package.json";
