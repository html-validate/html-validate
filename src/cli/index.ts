import { CLI } from "./cli";
export default CLI;
