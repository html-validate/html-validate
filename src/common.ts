/* used when calling require('htmlvalidate'); */

export { HtmlValidate } from "./htmlvalidate";
export { type AttributeData } from "./parser";
export {
	type ConfigData,
	type LoadedPlugin,
	type ResolvedConfigData,
	type Resolver,
	type ResolverOptions,
	type RuleConfig,
	type RuleOptions,
	type RuleSeverity,
	type StaticResolver,
	type StaticResolverMap,
	type TransformMap,
	type TransformerEntry,
	Config,
	ConfigError,
	ConfigLoader,
	ResolvedConfig,
	Severity,
	configPresets,
	defineConfig,
	staticResolver,
} from "./config";
export { StaticConfigLoader } from "./config/loaders/static";
export {
	type CSSStyleDeclaration,
	type DOMInternalID,
	type DOMNodeCache,
	Attribute,
	DOMNode,
	DOMTokenList,
	DOMTree,
	DynamicValue,
	HtmlElement,
	NodeClosed,
	NodeType,
	TextNode,
} from "./dom";
export { type EventDump, type RuleBlocker, type TokenDump } from "./engine";
export { NestedError, SchemaValidationError, UserError, WrappedError } from "./error";
export * from "./event";
export { version } from "./generated/package";
export {
	type AttrNameToken,
	type AttrValueToken,
	type BaseToken,
	type CommentToken,
	type ConditionalToken,
	type DirectiveToken,
	type DoctypeCloseToken,
	type DoctypeOpenToken,
	type DoctypeValueToken,
	type EOFToken,
	type ScriptToken,
	type StyleToken,
	type TagCloseToken,
	type TagOpenToken,
	type TemplatingToken,
	type TextToken,
	type Token,
	type TokenStream,
	type TokenType,
	type UnicodeBOMToken,
	type WhitespaceToken,
} from "./lexer";
export {
	type CategoryOrTag,
	type DeprecatedElement,
	type FormAssociated,
	type HtmlElementLike,
	type MetaAria,
	type MetaAttribute,
	type MetaAttributeAllowedCallback,
	type MetaData,
	type MetaDataTable,
	type MetaElement,
	type MetaFocusableCallback,
	type MetaImplicitRoleCallback,
	type MetaLookupableProperty,
	type MetadataHelper,
	type NormalizedMetaAria,
	type Permitted,
	type PermittedAttribute,
	type PermittedEntry,
	type PermittedGroup,
	type PermittedOrder,
	type PropertyExpression,
	type RequiredAncestors,
	type RequiredContent,
	MetaCopyableProperty,
	MetaTable,
	TextContent,
	Validator,
	defineMetadata,
	metadataHelper,
} from "./meta";
export {
	type ErrorDescriptor,
	type RuleConstructor,
	type RuleDocumentation,
	type SchemaObject,
	Rule,
} from "./rule";
export {
	type IncludeExcludeOptions,
	type TextClassificationOptions,
	ariaNaming,
	TextClassification,
	classifyNodeText,
	keywordPatternMatcher,
} from "./rules/helper";
export {
	type Location,
	type ProcessAttributeCallback,
	type ProcessElementCallback,
	type ProcessElementContext,
	type Source,
	type SourceHooks,
	sliceLocation,
} from "./context";
export { type DeferredMessage, type Message, type Report, type Result, Reporter } from "./reporter";
export { type TransformContext, type Transformer } from "./transform";
export { type Plugin, type SchemaValidationPatch, definePlugin } from "./plugin";
export { Parser } from "./parser";
export { type CompatibilityOptions, ruleExists } from "./utils";
