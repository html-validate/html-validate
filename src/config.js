'use strict';

const voidElements = [
	'area',
	'base',
	'br',
	'col',
	'embed',
	'hr',
	'img',
	'input',
	'keygen',
	'link',
	'menuitem',
	'meta',
	'param',
	'source',
	'track',
	'wbr',
];

function deepMerge(dst, src){
	for ( let key of Object.keys(src) ){
		if ( dst.hasOwnProperty(key) && typeof(dst[key]) === 'object' && typeof(src[key]) === 'object' ){
			deepMerge(dst[key], src[key]);
		} else {
			dst[key] = src[key];
		}
	}
	return dst;
}

class Config {
	constructor(options){
		this.config = {};
		this.loadDefaults();
		this.merge(options || {});
	}

	loadDefaults(){
		this.config = {
			html: {
				voidElements,
			},
			rules: {},
		};
	}

	merge(config){
		deepMerge(this.config, config);
	}

	get(){
		return Object.assign({}, this.config);
	}

	getRules(){
		return Object.assign({}, this.config.rules || {});
	}
}

module.exports = Config;
