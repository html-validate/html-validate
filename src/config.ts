const path = require('path');

const voidElements = [
	'area',
	'base',
	'br',
	'col',
	'embed',
	'hr',
	'img',
	'input',
	'keygen',
	'link',
	'menuitem',
	'meta',
	'param',
	'source',
	'track',
	'wbr',
];

const recommended = {
	rules: {
		'attr-quotes': 'error',
		'button-type': 'error',
		'close-attr': 'error',
		'close-order': 'error',
		'no-trailing-whitespace': 'error',
	},
};

function deepMerge(dst: any, src: any){
	for (const key of Object.keys(src)){
		if (dst.hasOwnProperty(key) && typeof dst[key] === 'object' && typeof src[key] === 'object'){
			deepMerge(dst[key], src[key]);
		} else {
			dst[key] = src[key];
		}
	}
	return dst;
}

const parseSeverityLut = [
	'disable',
	'warn',
	'error',
];

function parseSeverity(value: string | number){
	if (typeof value === 'number'){
		return value;
	} else {
		return parseSeverityLut.indexOf(value.toLowerCase());
	}
}

interface ConfigData {
	html: any;
	extends: Array<string>;
	rules: any;
}

class Config {
	config: ConfigData;

	public static readonly SEVERITY_DISABLED = 0;
	public static readonly SEVERITY_WARN = 1;
	public static readonly SEVERITY_ERROR = 2;

	static empty(): Config {
		return new Config();
	}

	static fromObject(options: Object): Config {
		return new Config(options);
	}

	static fromFile(filename: string): Config {
		if (filename === 'htmllint:recommended'){
			return Config.fromObject(recommended);
		}

		const json = require(filename);

		/* expand any relative paths */
		json.extends = (json.extends || []).map(function(ref: string){
			return Config.expandRelative(ref, path.dirname(filename));
		});

		return new Config(json);
	}

	constructor(options?: any){
		this.loadDefaults();
		this.merge(options || {});

		/* process and extended configs */
		const self = this;
		this.config.extends.forEach(function(ref: string){
			const base = Config.fromFile(ref);
			self.config = base.merge(self.config);
		});
	}

	private loadDefaults(){
		this.config = {
			html: {
				voidElements,
			},
			extends: [],
			rules: {},
		};
	}

	static expandRelative(src: string, currentPath: string): string {
		if (src[0] === '.'){
			return path.normalize(`${currentPath}/${src}`);
		}
		return src;
	}

	private merge(config: Object): ConfigData {
		deepMerge(this.config, config);
		return this.config;
	}

	get(): ConfigData {
		return Object.assign({}, this.config);
	}

	getRules(){
		const rules = Object.assign({}, this.config.rules || {});
		for (const name in rules){
			let options = rules[name];
			if (!Array.isArray(options)){
				options = [options, {}];
			} else if (options.length === 1){
				options.push({});
			}
			options[0] = parseSeverity(options[0]);
			rules[name] = options;
		}
		return rules;
	}

	isVoidElement(tagName: string): boolean {
		return this.config.html.voidElements.indexOf(tagName.toLowerCase()) !== -1;
	}
}

export default Config;
