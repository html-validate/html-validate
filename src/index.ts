/* entrypoint for nodejs build */

export * from "./common";
export { type CLIOptions, type ExpandOptions, type InitResult, CLI } from "./cli";
export {
	type FileSystemConfigLoaderOptions,
	FileSystemConfigLoader,
} from "./config/loaders/file-system";
export { type FSLike, type NodeJSResolver, nodejsResolver } from "./config/resolver/nodejs";
export {
	type AvailableFormatters,
	type Formatter,
	getFormatter as formatterFactory,
} from "./formatters";
export { compatibilityCheck } from "./utils/compatibility-check.nodejs";
