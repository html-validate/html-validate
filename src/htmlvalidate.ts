import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
