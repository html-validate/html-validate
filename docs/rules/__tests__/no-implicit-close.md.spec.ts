import { HtmlValidate } from "../../../src/htmlvalidate";

const markup: Record<string, string> = {};
markup["parent"] = `<ul>
    <li>foo
    <li>bar
    <li>baz
</ul>`;
markup["siblings"] = `<p>lorem ipsum
<p>dolor sit amet`;
markup["adjacent"] = `<p>
    <div>lorem ipsum</div>
</p>`;
markup["correct-list"] = `<ul>
     <li>foo</li>
     <li>bar</li>
     <li>baz</li>
</ul>`;
markup["correct-paragraph"] = `<p>lorem ipsum</p>
<p>dolor sit amet</p>`;

describe("docs/rules/no-implicit-close.md", () => {
	it("inline validation: parent", async () => {
		expect.assertions(1);
		const htmlvalidate = new HtmlValidate({"rules":{"no-implicit-close":"error"}});
		const report = await htmlvalidate.validateString(markup["parent"]);
		expect(report.results).toMatchSnapshot();
	});
	it("inline validation: siblings", async () => {
		expect.assertions(1);
		const htmlvalidate = new HtmlValidate({"rules":{"no-implicit-close":"error"}});
		const report = await htmlvalidate.validateString(markup["siblings"]);
		expect(report.results).toMatchSnapshot();
	});
	it("inline validation: adjacent", async () => {
		expect.assertions(1);
		const htmlvalidate = new HtmlValidate({"rules":{"no-implicit-close":"error"}});
		const report = await htmlvalidate.validateString(markup["adjacent"]);
		expect(report.results).toMatchSnapshot();
	});
	it("inline validation: correct-list", async () => {
		expect.assertions(1);
		const htmlvalidate = new HtmlValidate({"rules":{"no-implicit-close":"error"}});
		const report = await htmlvalidate.validateString(markup["correct-list"]);
		expect(report.results).toMatchSnapshot();
	});
	it("inline validation: correct-paragraph", async () => {
		expect.assertions(1);
		const htmlvalidate = new HtmlValidate({"rules":{"no-implicit-close":"error"}});
		const report = await htmlvalidate.validateString(markup["correct-paragraph"]);
		expect(report.results).toMatchSnapshot();
	});
});
