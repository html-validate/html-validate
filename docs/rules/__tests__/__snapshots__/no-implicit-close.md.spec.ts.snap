// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs/rules/no-implicit-close.md inline validation: adjacent 1`] = `
Array [
  Object {
    "errorCount": 1,
    "filePath": "inline",
    "messages": Array [
      Object {
        "column": 2,
        "context": undefined,
        "line": 1,
        "message": "Element <p> is implicitly closed by adjacent <div>",
        "offset": 1,
        "ruleId": "no-implicit-close",
        "selector": "p",
        "severity": 2,
        "size": 1,
      },
    ],
    "source": "<p>
    <div>lorem ipsum</div>
</p>",
    "warningCount": 0,
  },
]
`;

exports[`docs/rules/no-implicit-close.md inline validation: correct-list 1`] = `Array []`;

exports[`docs/rules/no-implicit-close.md inline validation: correct-paragraph 1`] = `Array []`;

exports[`docs/rules/no-implicit-close.md inline validation: parent 1`] = `
Array [
  Object {
    "errorCount": 3,
    "filePath": "inline",
    "messages": Array [
      Object {
        "column": 6,
        "context": undefined,
        "line": 2,
        "message": "Element <li> is implicitly closed by sibling",
        "offset": 10,
        "ruleId": "no-implicit-close",
        "selector": "ul > li:nth-child(1)",
        "severity": 2,
        "size": 2,
      },
      Object {
        "column": 6,
        "context": undefined,
        "line": 3,
        "message": "Element <li> is implicitly closed by sibling",
        "offset": 22,
        "ruleId": "no-implicit-close",
        "selector": "ul > li:nth-child(2)",
        "severity": 2,
        "size": 2,
      },
      Object {
        "column": 6,
        "context": undefined,
        "line": 4,
        "message": "Element <li> is implicitly closed by parent </ul>",
        "offset": 34,
        "ruleId": "no-implicit-close",
        "selector": "ul > li:nth-child(3)",
        "severity": 2,
        "size": 2,
      },
    ],
    "source": "<ul>
    <li>foo
    <li>bar
    <li>baz
</ul>",
    "warningCount": 0,
  },
]
`;

exports[`docs/rules/no-implicit-close.md inline validation: siblings 1`] = `
Array [
  Object {
    "errorCount": 1,
    "filePath": "inline",
    "messages": Array [
      Object {
        "column": 2,
        "context": undefined,
        "line": 1,
        "message": "Element <p> is implicitly closed by sibling",
        "offset": 1,
        "ruleId": "no-implicit-close",
        "selector": "p:nth-child(1)",
        "severity": 2,
        "size": 1,
      },
    ],
    "source": "<p>lorem ipsum
<p>dolor sit amet",
    "warningCount": 0,
  },
]
`;
